File: backend/src/main/java/com/trako/entities/RecurringTransaction.java
package com.trako.entities;

import com.trako.enums.Frequency;
import com.trako.enums.TransactionType;
import com.trako.enums.TransactionTypeConverter;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Positive;
import jakarta.validation.constraints.Size;
import jakarta.validation.constraints.Pattern;
import lombok.Getter;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.util.Date;

@Entity
@Table(name = "recurring_transactions")
@Getter
@Setter
public class RecurringTransaction {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private Long id;

    @Column(name = "user_id", length = 36, nullable = false)
    private String userId;

    @NotBlank
    @Size(max = 128)
    @Column(name = "name", length = 128)
    private String name;

    @Column(name = "amount", insertable = false, updatable = false)
    private Double amount;

    @NotBlank
    @Size(min = 3, max = 3)
    @Pattern(regexp = "[A-Z]{3}", message = "originalCurrency must be a 3-letter uppercase currency code")
    @Column(name = "original_currency", length = 3, nullable = false)
    private String originalCurrency;

    @NotNull
    @Positive
    @Column(name = "original_amount", nullable = false)
    private Double originalAmount;

    @Column(name = "exchange_rate")
    @Positive
    private Double exchangeRate;

    @NotNull
    @Positive
    @Column(name = "account_id")
    private Long accountId;

    @NotNull
    @Positive
    @Column(name = "category_id")
    private Long categoryId;

    @Column(name = "to_account_id")
    @Positive
    private Long toAccountId;

    @NotNull
    @Column(name = "transaction_type")
    @Convert(converter = TransactionTypeConverter.class)
    private TransactionType transactionType;

    @NotNull
    @Enumerated(EnumType.STRING)
    @Column(name = "frequency", length = 20)
    private Frequency frequency;

    @NotNull
    @Temporal(TemporalType.TIMESTAMP)
    @Column(name = "start_date")
    private Date startDate;

    @NotNull
    @Temporal(TemporalType.TIMESTAMP)
    @Column(name = "next_run_date")
    private Date nextRunDate;

    @Temporal(TemporalType.TIMESTAMP)
    @Column(name = "end_date")
    private Date endDate;

    @Column(name = "is_active")
    private Boolean isActive = true;

    @Temporal(TemporalType.TIMESTAMP)
    @Column(name = "last_run_date")
    private Date lastRunDate;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private Date createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private Date updatedAt;
}

File: frontend/lib/models/recurring_transaction.dart
import 'package:tracko/Utils/enums.dart';

enum Frequency {
  DAILY,
  WEEKLY,
  MONTHLY,
  YEARLY;

  String toJson() => name;
  static Frequency fromJson(String json) => values.byName(json);
}

class RecurringTransaction {
  int? id;
  String userId = '';
  String name = '';
  double? amount;
  int accountId = 0;
  int categoryId = 0;
  int? toAccountId;
  int transactionType = TransactionType.DEBIT;
  Frequency frequency = Frequency.MONTHLY;
  DateTime startDate = DateTime.now();
  DateTime nextRunDate = DateTime.now();
  DateTime? endDate;
  bool isActive = true;
  DateTime? lastRunDate;
  DateTime? createdAt;
  DateTime? updatedAt;

  // Currency Support
  String? originalCurrency;
  double? originalAmount;
  double? exchangeRate;

  RecurringTransaction();

  factory RecurringTransaction.fromJson(Map<String, dynamic> json) {
    final rt = RecurringTransaction();
    rt.id = json['id'];
    rt.userId = json['userId'] ?? '';
    rt.name = json['name'] ?? '';
    rt.amount = (json['amount'] as num?)?.toDouble();
    rt.accountId = json['accountId'] ?? 0;
    rt.categoryId = json['categoryId'] ?? 0;
    rt.toAccountId = json['toAccountId'];
    rt.transactionType = json['transactionType'] ?? TransactionType.DEBIT;
    rt.frequency = Frequency.values.firstWhere(
        (e) => e.name == (json['frequency'] as String),
        orElse: () => Frequency.MONTHLY);
    rt.startDate = DateTime.parse(json['startDate']);
    rt.nextRunDate = DateTime.parse(json['nextRunDate']);
    rt.endDate =
        json['endDate'] != null ? DateTime.parse(json['endDate']) : null;
    rt.isActive = json['isActive'] ?? true;
    rt.lastRunDate = json['lastRunDate'] != null
        ? DateTime.parse(json['lastRunDate'])
        : null;
    rt.createdAt =
        json['createdAt'] != null ? DateTime.parse(json['createdAt']) : null;
    rt.updatedAt =
        json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : null;

    rt.originalCurrency = json['originalCurrency'];
    rt.originalAmount = (json['originalAmount'] as num?)?.toDouble();
    rt.exchangeRate = (json['exchangeRate'] as num?)?.toDouble();

    return rt;
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'userId': userId,
      'name': name,
      'amount': amount,
      'accountId': accountId,
      'categoryId': categoryId,
      'toAccountId': toAccountId,
      'transactionType': transactionType,
      'frequency': frequency.name,
      'startDate': startDate.toIso8601String(),
      'nextRunDate': nextRunDate.toIso8601String(),
      'endDate': endDate?.toIso8601String(),
      'isActive': isActive,
      'lastRunDate': lastRunDate?.toIso8601String(),
      'originalCurrency': originalCurrency,
      'originalAmount': originalAmount,
      'exchangeRate': exchangeRate,
    };
  }
}
